src/mount_efs/__init__.py

```diff
@@ -578,6 +574,72 @@ def _validate_replacement_field_count(format_str, expected_ct):
     return dns_name
 
 
+def match_device(config, device):
+    """
+    Return the EFS id and the remote path to mount.
+
+    :param config: the current configuration
+    :param device: the device descriptor, separating an EFS id or a DNS name and the remote path to mount by a colon
+    :return: a two element tuple of the EFS id and the remote path to mount
+    """
+
+    # The device descriptor as specified separates the remote filesystem by the path to mount by a colon. Since colons
+    # are not allowed in either an EFS id or in a domain name, we can left-split once. (If left-splitting fails, the
+    # user didn't specify a path and we use '/' as the default.)
+    try:
+        remote, path = device.split(':', 1)
+    except ValueError:
+        remote = device
+        path = '/'
+
+    # The simplest case is that the remote is already the EFS id. If so, we return it as is.
+    if FS_ID_RE.match(remote):
+        return remote, path
+
+    # If the user did not specify an EFS id, we first check for the special case where the user supplied us with the
+    # FQDN of the EFS.
+    efs_fqdn_match = EFS_FQDN_RE.match(remote)
+    if efs_fqdn_match:
+        fs_id = efs_fqdn_match.group('fs_id')
+        expected_dns_name = get_dns_name(config, fs_id)
+        if remote == expected_dns_name:
+            return fs_id, path
+        else:
+            fatal_error(
+                'Fully qualified EFS domain name specified "%s", but it didn\'t match the expected value "%s"'
+                % (remote, expected_dns_name),
+                'EFS FQDN "%s" didn\'t match expected "%s"' % (remote, expected_dns_name)
+            )
+
+    # For the final case we now assume that the user specified a DNS resolvable name with a CNAME record that points to
+    # a valid EFS FQDN. To verify this, we'll use `socket.gethostbyname_ex()` which returns an alias-list for, among
+    # other things, a CNAME.
+    try:
+        primary, secondaries, _ = socket.gethostbyname_ex(remote)
+        hostnames = filter(lambda e: e is not None, [primary] + secondaries)
+    except socket.gaierror:
+        fatal_error(
+            'Failed to resolve "%s" - check that the specified DNS name is a CNAME record resolving to a valid EFS DNS '
+            'name' % remote,
+            'Failed to resolve "%s"' % remote
+        )
+
+    if not hostnames:
+        fatal_error(
+            'The specified domain name "%s" returned no entries, where at least one was expected' % remote
+        )
+
+    for hostname in hostnames:
+        efs_fqdn_match = EFS_FQDN_RE.match(hostname)
+        if efs_fqdn_match:
+            fs_id = efs_fqdn_match.group('fs_id')
+            expected_dns_name = get_dns_name(config, fs_id)
+            if hostname == expected_dns_name:
```

Comment:
Why do this check instead of just returning `fs_id, path` as soon as something matches the regex? Can you explain why you're doing the 2nd check?

Reply:
I mainly did it to make sure that the [DNS name that will ultimately be used for mounting](https://github.com/aws/efs-utils/blob/87af3b6de45b3bbae143c649db58bc726d2fafed/src/mount_efs/__init__.py#L614) is identical to the DNS name that the CNAME resolved to.

I was imagining the following case:

* The user specified their CNAME `efs.example.com` to resolve to `fs-deadbeef.efs.us-west-1.amazonaws.com`.
* The user is now trying to mount this filesystem in an EC2 that he launched in a different region **`us-east-1`** by CNAME `efs.example.com`.

If we didn't perform this second comparison, we would silently "correct" the EFS DNS name to `fs-deadbeef.efs.us-east-1.amazonaws.com` which in turn would obviously fail due to a non-existant EFS.

Reply:
I see now. Give we had this back and forth here can you please summarize this in a comment in the code? Something like this would be good enough:
```
# check that the DNS name of the mount target matches exactly the DNS name the CNAME resolves to
```

Reply:
Good point, [added](https://github.com/aws/efs-utils/pull/14/commits/636da3503b8d2cd6c4a9aef492472ede9bea092d)!